editor/src/messages/portfolio/document/utility_types/network_interface/resolved_types.rs

```diff
@@ -0,0 +1,305 @@
+				}
+			}
+			OutputConnector::Import(import_index) => {
+				let Some((encapsulating_node, encapsulating_path)) = network_path.split_last() else {
+					return TypeSource::Error("Cannot get import type in document network");
+				};
+				self.input_type(&InputConnector::node(*encapsulating_node, *import_index), encapsulating_path)
+			}
+		}
+	}
+	// The valid output types are all types that are valid for each downstream connection
+	pub fn valid_output_types(&mut self, output_connector: &OutputConnector, network_path: &[NodeId]) -> Result<Vec<Type>, String> {
+		let Some(outward_wires) = self.outward_wires(&network_path) else {
+			return Err("Could not get outward wires in valid_input_types".to_string());
+		};
+		let Some(inputs_from_import) = outward_wires.get(output_connector) else {
+			return Err("Could not get inputs from import in valid_input_types".to_string());
+		};
+
+		let intersection = inputs_from_import
+			.clone()
+			.iter()
+			.map(|input_connector| self.valid_input_types(input_connector, &network_path).into_iter().collect::<HashSet<_>>())
+			.fold(None, |acc: Option<HashSet<Type>>, set| match acc {
+				Some(acc_set) => Some(acc_set.intersection(&set).cloned().collect()),
+				None => Some(set),
+			})
+			.unwrap_or_default();
+
+		Ok(intersection.into_iter().collect::<Vec<_>>())
+	}
+
+	/// Performs a downstream iteration from an input connector to the next protonode and selects a random type from its implementations.
+	pub fn random_type_for_connector(&mut self, input_connector: &InputConnector, network_path: &[NodeId]) -> Option<Type> {
+		match input_connector {
+			InputConnector::Node { node_id, input_index } => {
+				let Some(implementation) = self.implementation(node_id, network_path) else {
+					log::error!("Could not get node {node_id} in random_downstream_protonode_from_connector");
+					return None;
+				};
+				match implementation {
+					DocumentNodeImplementation::Network(_) => {
+						let Some(outward_wires) = self.outward_wires(&network_path) else {
+							log::error!("Could not get outward wires in random_downstream_protonode_from_connector");
+							return None;
+						};
+						let Some(inputs_from_import) = outward_wires.get(&OutputConnector::Import(*input_index)) else {
+							log::error!("Could not get inputs from import in valid_input_types");
+							return None;
+						};
+						let Some(first_input) = inputs_from_import.first().cloned() else {
+							return None;
+						};
+						self.random_type_for_connector(&first_input, &[network_path, &[*node_id]].concat())
+					}
+					DocumentNodeImplementation::ProtoNode(proto_node_identifier) => {
+						let Some(implementations) = NODE_REGISTRY.get(proto_node_identifier) else {
+							log::error!("Protonode {proto_node_identifier:?} not found in registry");
+							return None;
+						};
+						implementations.keys().next().and_then(|node_io| node_io.inputs.get(input_connector.input_index())).cloned()
```

Comment:
Couldn't we just sort the keys to make this function deterministic? The output will still not be as useful but having it nondeterministic just seems unnecessary